File: assignment 10/python/programs/5.py
class Node:
    def __init__(self, key):
        self.left = None
        self.right = None
        self.value = key


class BinarySearchTree:
    def __init__(self):
        self.root = None

    def insert(self, value):
        self.root = self._insert_recursive(self.root, value)

    def _insert_recursive(self, current_node, value):
        if current_node is None:
            return Node(value)
        if value < current_node.value:
            current_node.left = self._insert_recursive(current_node.left, value)
        elif value > current_node.value:
            current_node.right = self._insert_recursive(current_node.right, value)
        return current_node

    def inorder_traversal(self):
        return self._inorder_traversal_recursive(self.root, [])

    def _inorder_traversal_recursive(self, current_node, result):
        if current_node:
            result = self._inorder_traversal_recursive(current_node.left, result)
            result.append(current_node.value)
            result = self._inorder_traversal_recursive(current_node.right, result)
        return result


def sort_numbers_using_bst(numbers):
    tree = BinarySearchTree()
    for number in numbers:
        tree.insert(number)
    return tree.inorder_traversal()


# Example usage:
if __name__ == "__main__":
    numbers = [12, 4, 7, 2, 14, 9, 3, 17]
    sorted_numbers = sort_numbers_using_bst(numbers)
    print("Sorted Numbers:", sorted_numbers)

File: assignment 10/c/programs/5.c
#include <stdio.h>
#include <stdlib.h>

// Define the structure for a binary search tree node
struct Node
{
    int data;
    struct Node *left;
    struct Node *right;
};

// Function to create a new node
struct Node *createNode(int data)
{
    struct Node *newNode = (struct Node *)malloc(sizeof(struct Node));
    newNode->data = data;
    newNode->left = NULL;
    newNode->right = NULL;
    return newNode;
}

// Function to insert a node into the binary search tree
struct Node *insert(struct Node *root, int data)
{
    if (root == NULL)
    {
        return createNode(data);
    }

    if (data < root->data)
    {
        root->left = insert(root->left, data);
    }
    else if (data > root->data)
    {
        root->right = insert(root->right, data);
    }

    return root;
}

// Function to perform in-order traversal of the binary search tree
void inOrderTraversal(struct Node *root)
{
    if (root != NULL)
    {
        inOrderTraversal(root->left);
        printf("%d ", root->data);
        inOrderTraversal(root->right);
    }
}

int main()
{
    struct Node *root = NULL;

    // Input numbers to be sorted
    int numbers[] = {50, 30, 20, 40, 70, 60, 80};
    int numCount = sizeof(numbers) / sizeof(numbers[0]);

    // Insert numbers into the binary search tree
    for (int i = 0; i < numCount; i++)
    {
        root = insert(root, numbers[i]);
    }

    // Perform in-order traversal to display the sorted numbers
    printf("Sorted numbers: ");
    inOrderTraversal(root);
    printf("\n");

    return 0;
}
